### PR TITLE
Migrate taxonomy settings to taxonomy specific namespace

### PR DESCRIPTION
**Pull request by davisshaver**
Identified by @dlh01, right now post types and taxonomies share the same namespace. This patch handles migrating legacy data to a new taxonomy specific namespace. https://github.com/alleyinteractive/wp-seo/pull/72#issuecomment-287625525

### CODE REVIEW

File: php/class-wp-seo-settings.php [hunk 1]
```diff
@@ -162,6 +171,32 @@ public function set_properties() {
 		 * @param  array Associative array of setting names and values.
 		 */
 		$this->default_options = apply_filters( 'wp_seo_default_options', array( 'post_types' => array_keys( $this->single_post_types ), 'taxonomies' => array_keys( $this->taxonomies ) ) );
+
+		/**
+		 * Checks database to see if taxonomy migration has been run.
+		 *
+		 * @since v0.13.0
+		 */
+		$this->taxonomy_migration = get_option( $this::SLUG . '-taxonomy-migration', false );
```
Comment (dlh01) [inline php/class-wp-seo-settings.php]: A new option seems a little heavy, especially if we update something else about the option structure again later. 

Could we store the status in our existing option?

Thinking out loud, we could add a top-level `internal` key to the option array and make it clear that the data are for core plugin use only. Then our check could be like `if ( empty( $option['internal']['migrated_taxonomies_to_taxonomy_key'] ) )...`.

File: php/general-functions.php [hunk 2]
```diff
@@ -42,7 +42,7 @@ function wp_seo_get_key( $query = false ) {
 		$key = 'archive_author';
 	} elseif ( $query->is_category() || $query->is_tag() || $query->is_tax() ) {
 		$taxonomy = get_queried_object()->taxonomy;
-		$key = "archive_{$taxonomy}";
+		$key = "taxonomy_{$taxonomy}";
```
Comment (dlh01) [inline php/general-functions.php]: As far as I know, there is no "official" method in the plugin for getting data from the option other than to request it by its array key. If so, changing the key outright seems like a backwards-compatibility break. If so, do you think it's worth accounting for?

File: php/class-wp-seo-settings.php (hunk 1 above)
Comment (davisshaver) [inline php/class-wp-seo-settings.php]: Great idea!

File: php/general-functions.php (hunk 2 above)
Comment (davisshaver) [inline php/general-functions.php]: Definitely, thank you for pointing that out @dlh01

**Comment by davisshaver** [comment]
@dlh01: I've taken another stab at this logic. The commit tree got a little messy so let me know what you think about rebasing (I was fixing PHPCS errors from the updated rulesets).

My description of the logic is as follows...

- Add `internal->archive_to_taxonomy_migration` as `true` to the default options (used if there hadn't been an option value saved previously).
- If this value isn't set/set as true, the current key will be used for taxonomies (`archive_`). Otherwise we'll use the new key (`taxonomy_`).
- The value gets set on save when default options is merged inside `sanitize_options`.

**Comment by dlh01** [comment]
@davisshaver Sorry for taking so long to respond. I think I'm on board with the general plan, but I'm not sure I follow what happens for someone who is currently calling `wp_seo_settings()->get_option( 'archive_mytaxonomy_* )`. I might be missing it -- Is support provided in that case?

Another potential backwards-compatibility concern I didn't think about before is that `$key` is passed to filters such as `wp_seo_title_tag_format`. Once the settings migration runs, the passed value will change.

**Comment by davisshaver** [comment]
No worries @dlh01!

> @davisshaver Sorry for taking so long to respond. I think I'm on board with the general plan, but I'm not sure I follow what happens for someone who is currently calling wp_seo_settings()->get_option( 'archive_mytaxonomy_* ). I might be missing it -- Is support provided in that case?

Yes. `wp_seo_settings()->get_option()` merges the default options with database values for the `wp-seo` option. If that option has not been migrated, the default value will be set – see https://github.com/alleyinteractive/wp-seo/blob/archive-to-taxonomy/php/class-wp-seo-settings.php#L211-L213. Note that `get_option()` does not merge values, so within `wp_seo_settings()->has_taxonomy_migration_run()` we set `false` as the default value.

To the more general question, I don't think there should be any implementations of `wp_seo_settings()->get_option( 'archive_mytaxonomy_* )` anywhere outside of the plugin, correct? The logic above should mean that before the migration is run, the key `archive_mytaxonomy_*` will be checked, and after the migration is run, the key `taxonomy_mytaxonomy_*` will be checked.

> Another potential backwards-compatibility concern I didn't think about before is that $key is passed to filters such as wp_seo_title_tag_format. Once the settings migration runs, the passed value will change.

Hm yes... This does seem like a breaking change. Same goes for `wp_seo_meta_*_format` filters. Making these filters backward compatibility seems like it'd add more complexity than it's worth. Maybe on activation we check with `has_filter()` and show a change notice to admins?

**Comment by dlh01** [comment]
> I don't think there should be any implementations of wp_seo_settings()->get_option( 'archive_mytaxonomy_* ) anywhere outside of the plugin, correct?

This is what I don't know. I guess it's probably unusual, but if it was out there, would it break?

> Maybe on activation we check with has_filter() and show a change notice to admins?

My initial response leans against this approach. I suspect it would reach a lot of people who wouldn't be able to fix the code; we wouldn't be able to tell whether the filters we detected needed to be changed; perhaps above all, the damage would already have been done.

**Comment by davisshaver** [comment]
> This is what I don't know. I guess it's probably unusual, but if it was out there, would it break?

Yes I think it would break functionality.

> My initial response leans against this approach. I suspect it would reach a lot of people who wouldn't be able to fix the code; we wouldn't be able to tell whether the filters we detected needed to be changed; perhaps above all, the damage would already have been done.

We could send an email to site admin. Functionality would break but the damage could certainly be reversed. 

How would you like to take it from here?

**Comment by dlh01** [comment]
I don't know. My guess is that breaking backcompat is more likely to affect people than fixing the bug. The bug has existed in the plugin almost since the dawn of time and has gone unreported, to my knowledge. So from a utilitarian perspective, I think I'm reluctant to fix it if compatibility can't be preserved. On the other hand, I have only that guess, not data. The population of both groups of people could be zero. What do you think?

**Comment by davisshaver** [comment]
@dlh01 Hm.... Well first off, do you agree with the assessment that there is _not_ a means to address the bug without breaking backward compatibility?

I see arguments for both merging the fix and not merging. It's hard to check all the possible strings that might be used for a custom taxonomy filter. We don't know if users outside Alley have implemented these filters. 

If we decide to "keep" the bug, maybe we would want to send admins an email on upgrade that we've identified this as an issue (or possibly only if they have a category and taxonomy that share the same name?).

**Pull request merged**